File: cycle2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        
        slow = head
        fast = head
        flag = False
        while fast is not None and fast.next is not None :
            slow = slow.next
            fast = fast.next.next
            
            if slow==fast:
                flag = True
                break
            
        if flag == False:
            return None
        
        fast = head
        
        while slow!=fast:
            slow = slow.next
            fast = fast.next
            
        return slow
        
# approach is to keep one fast and slow pointer. time complexity is O(2n) ~ O(n) and space complexity is O(1)

File: cycle2.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if(head==null){
            return null;
        }
        ListNode slow=head;
        ListNode fast=head;
        boolean hasCycle=false;
        while(fast!=null && fast.next!=null){
            slow=slow.next;
            fast=fast.next.next;
            if (slow==fast){
                hasCycle=true;
                break;
            }
        }
        if(!hasCycle){
            return null;
        }
        slow=head;
        while(fast!=slow){
            slow=slow.next;
            fast=fast.next;
        }
        return slow;
    }
}